### DISCUSSION TITLE
EROFS: read-only file system, unlink '/workspace/dist/next/BUILD_ID'

### DISCUSSION THREAD

**Discussion by rpyn** [vercel/next.js#36743] [Help]
I am building a Next.js app using the version `12.1.6` being deployed on Firebase Cloud Functions. In the local dev environment, everything is good. It works great, but after it's deployed to the server, it returns `Failed to load resource: the server responded with a status of 503 ()` to the client. There is an error message in the functions log saying:

```
Error: EROFS: read-only file system, unlink '/workspace/dist/next/BUILD_ID'] {
  errno: -30,
  code: 'EROFS',
  syscall: 'unlink',
  path: '/workspace/dist/next/BUILD_ID'
}
```

It seems that Next.js tries to do write operations in the node environment. 

Firebase Cloud Functions only offers the `/temp` directory where writable actions are allowed in the local disk, meaning that it does not accept any writable actions outside the `/temp` directory. Only read operation.

In another project using Next.js v9.3, it actually works so the error must have come out after newer versions than v9.3, I believe. Maybe this bug is peculiar only to Next.js apps being Cloud Functions, but I would like to know how to avoid this error in an app using the latest Next version and Firebase Cloud Functions? 

**packag.json**

```
 "engines": {
    "node": "16"
  },
  "dependencies": {
    "firebase-admin": "^10.0.2",
    "firebase-functions": "^3.20.1",
    "next": "^12.1.6",
    "react": "17.0.2",
    "react-dom": "17.0.2"
  },
  "devDependencies": {
    "@types/node": "17.0.13",
    "@types/react": "17.0.38",
    "eslint": "8.8.0",
    "eslint-config-next": "12.0.9",
    "typescript": "4.5.5"
  }
```

**firebase.json**

```
{
  "firestore": {
    "rules": "firestore.rules",
    "indexes": "firestore.indexes.json"
  },
  "functions": {
    "source": ".",
    "predeploy": ["npm run build"],
    "runtime": "nodejs16"
  },
  "hosting": {
    "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
    "rewrites": [
      {
        "function": "nextHosting",
        "source": "**"
      }
    ]
  }
}
```

**Comment by rpyn** [comment]
As I inspected more on this, it seems the `fs.unlink` part at `~/node_modules/next/dist/compiled/webpack/bundle5.js`  is causing this error. Also, I confirm that the same javascript file do not exist in another project using v9.3, which is working without any issue. So maybe the newly added `webpack5` is triggering the error? 

Is there any way to prevent this bug in the newest version? Like flag?

```
const applyDiff = (fs, outputPath, dry, logger, diff, isKept, callback) => {
	const log = msg => {
		if (dry) {
			logger.info(msg);
		} else {
			logger.log(msg);
		}
	};
	/** @typedef {{ type: "check" | "unlink" | "rmdir", filename: string, parent: { remaining: number, job: Job } | undefined }} Job */
	/** @type {Job[]} */
	const jobs = Array.from(diff.keys(), filename => ({
		type: "check",
		filename,
		parent: undefined
	}));
	/** @type {Assets} */
	const keptAssets = new Map();
	processAsyncTree(
		jobs,
		10,
		({ type, filename, parent }, push, callback) => {
			const handleError = err => {
				if (err.code === "ENOENT") {
					log(`${filename} was removed during cleaning by something else`);
					handleParent();
					return callback();
				}
				return callback(err);
			};
			const handleParent = () => {
				if (parent && --parent.remaining === 0) push(parent.job);
			};
			const path = join(fs, outputPath, filename);
			switch (type) {
				case "check":
					if (isKept(filename)) {
						keptAssets.set(filename, 0);
						// do not decrement parent entry as we don't want to delete the parent
						log(`${filename} will be kept`);
						return process.nextTick(callback);
					}
					doStat(fs, path, (err, stats) => {
						if (err) return handleError(err);
						if (!stats.isDirectory()) {
							push({
								type: "unlink",
								filename,
								parent
							});
							return callback();
						}
						fs.readdir(path, (err, entries) => {
							if (err) return handleError(err);
							/** @type {Job} */
							const deleteJob = {
								type: "rmdir",
								filename,
								parent
							};
							if (entries.length === 0) {
								push(deleteJob);
							} else {
								const parentToken = {
									remaining: entries.length,
									job: deleteJob
								};
								for (const entry of entries) {
									const file = /** @type {string} */ (entry);
									if (file.startsWith(".")) {
										log(
											`${filename} will be kept (dot-files will never be removed)`
										);
										continue;
									}
									push({
										type: "check",
										filename: `${filename}/${file}`,
										parent: parentToken
									});
								}
							}
							return callback();
						});
					});
					break;
				case "rmdir":
					log(`${filename} will be removed`);
					if (dry) {
						handleParent();
						return process.nextTick(callback);
					}
					if (!fs.rmdir) {
						logger.warn(
							`${filename} can't be removed because output file system doesn't support removing directories (rmdir)`
						);
						return process.nextTick(callback);
					}
					fs.rmdir(path, err => {
						if (err) return handleError(err);
						handleParent();
						callback();
					});
					break;
				case "unlink":
					log(`${filename} will be removed`);
					if (dry) {
						handleParent();
						return process.nextTick(callback);
					}
					if (!fs.unlink) {
						logger.warn(
							`${filename} can't be removed because output file system doesn't support removing files (rmdir)`
						);
						return process.nextTick(callback);
					}
					fs.unlink(path, err => {
						if (err) return handleError(err);
						handleParent();
						callback();
					});
					break;
			}
		},
		err => {
			if (err) return callback(err);
			callback(undefined, keptAssets);
		}
	);
};
```

The app crashes because it is being executed outside the `/temp` directory of Firebase Cloud Functions.
```
~
fs.unlink(path, err => {
   if (err) return handleError(err);
        handleParent();
        callback();
});
~
```

As a workaround, I tried to use webpack4 instead of webpack5 by updating the `next.config.js` file. I followed the [doc](https://nextjs.org/docs/messages/webpack5) but it ended up forcing me to use webpack5.

**Comment by briandonahue** [comment]
I'm seeing this issue as well.  Did you ever find a solution?

**Reply by rpyn** [reply]
Nope. Our project is still using the version of 9.3.6 because of this strange issue.

@briandonahue 
Did you use GCP cloud functions as well? Is there any difference between your setting/versions and my mine?

**Reply by briandonahue** [reply]
@rpyn I was running into so many issues and complexities trying to get Next.js running smoothly with Firebase Hosting and Functions, I eventually decided to go full static with Next, deploy to FB hosting, and separately deploy FB Functions as the API layer with no Next SSR.  I may revisit it, but just even getting a single repo set up and deploying reliably with Functions and Next on FB Hosting was a nightmare.  Things like the emulators read/handle env vars differently than when you deploy, and all kinds of gotchas.  I gave up.

**Comment by sdjnes** [comment]
Facing a similar issue deploying to Google App Engine. We are migrating a project and want to leverage the latest version of Next (including SWC).

**Comment by ethanhe42** [comment]
tldr use cloud function v2
https://yihui.dev/firebase-hosting-with-nextjs-and-cloud-functions

**Reply by briandonahue** [reply]
Great writeup.  We eneded up going with a 100% static site, but would like to give this a shot in the future.
